pkg/services/ngalert/state/manager_test.go

```diff
@@ -282,2076 +280,999 @@ func TestDashboardAnnotations(t *testing.T) {
 }
 
 func TestProcessEvalResults(t *testing.T) {
-	evaluationTime, err := time.Parse("2006-01-02", "2021-03-25")
-	if err != nil {
-		t.Fatalf("error parsing date format: %s", err.Error())
-	}
 	evaluationDuration := 10 * time.Millisecond
+	evaluationInterval := 10 * time.Second
 
-	testCases := []struct {
+	t1 := time.Time{}.Add(evaluationInterval)
+
+	tn := func(n int) time.Time {
+		return t1.Add(time.Duration(n-1) * evaluationInterval)
+	}
+
+	t2 := tn(2)
+	t3 := tn(3)
+
+	baseRule := &models.AlertRule{
+		OrgID: 1,
+		Title: "test_title",
+		UID:   "test_alert_rule_uid",
+		Data: []models.AlertQuery{{
+			RefID:         "A",
+			DatasourceUID: "datasource_uid_1",
+		}, {
+			RefID:         "B",
+			DatasourceUID: expr.DatasourceType,
+		}},
+		NamespaceUID:    "test_namespace_uid",
+		Annotations:     map[string]string{"annotation": "test"},
+		Labels:          map[string]string{"label": "test"},
+		IntervalSeconds: int64(evaluationInterval.Seconds()),
+		NoDataState:     models.NoData,
+		ExecErrState:    models.ErrorErrState,
+	}
+
+	newEvaluation := func(evalTime time.Time, evalState eval.State) state.Evaluation {
+		return state.Evaluation{
+			EvaluationTime:  evalTime,
+			EvaluationState: evalState,
+			Values:          make(map[string]*float64),
+		}
+	}
+
+	baseRuleWith := func(mutators ...models.AlertRuleMutator) *models.AlertRule {
+		r := models.CopyRule(baseRule)
+		for _, mutator := range mutators {
+			mutator(r)
+		}
+		return r
+	}
+
+	newResult := func(mutators ...eval.ResultMutator) eval.Result {
+		r := eval.Result{
+			State:              eval.Normal,
+			EvaluationDuration: evaluationDuration,
+		}
+		for _, mutator := range mutators {
+			mutator(&r)
+		}
+		return r
+	}
+
+	labels1 := data.Labels{
+		"instance_label": "test-1",
+	}
+	labels2 := data.Labels{
+		"instance_label": "test-2",
+	}
+	systemLabels := data.Labels{
+		"system": "owned",
+	}
+	noDataLabels := data.Labels{
+		"datasource_uid": "1",
+		"ref_id":         "A",
+	}
+	labels := map[string]data.Labels{
+		"system + rule":           mergeLabels(baseRule.Labels, systemLabels),
+		"system + rule + labels1": mergeLabels(mergeLabels(labels1, baseRule.Labels), systemLabels),
+		"system + rule + labels2": mergeLabels(mergeLabels(labels2, baseRule.Labels), systemLabels),
+		"system + rule + no-data": mergeLabels(mergeLabels(noDataLabels, baseRule.Labels), systemLabels),
+	}
```

Comment:
I chose this approach to declaring labels of the states. I think it is much clearer what labels a state should have